clap_complete/tests/testsuite/dynamic.rs

```diff
@@ -1,34 +1,69 @@
 #![cfg(feature = "unstable-dynamic")]
 
+macro_rules! assert_completion {
+    ($cmd:expr, $input:expr, $(current_dir = $current_dir:expr,)? [$($comp:expr $(=> $help:expr)?),+]) => {
+        let mut args = vec![std::ffi::OsString::from($cmd.get_name())];
+        let arg_index;
+
+        if let Some((prior, after)) = $input.split_once("[TAB]") {
+            args.extend(prior.split_whitespace().map(From::from));
+            if prior.ends_with(char::is_whitespace) {
+                args.push(std::ffi::OsString::default())
+            }
+            arg_index = args.len() - 1;
+            // TODO when we can handle [TAB] inside of tokens, we need to update this
+            args.extend(after.split_whitespace().map(From::from));
+        } else {
+            args.extend($input.split_whitespace().map(From::from));
+            if $input.ends_with(char::is_whitespace) {
+                args.push(std::ffi::OsString::default())
+            }
+            arg_index = args.len() - 1;
+        }
+
+        let current_dir = None;
+        $(let current_dir = $current_dir;)?
+
+        let mut completions =
+            clap_complete::dynamic::complete(&mut $cmd, args, arg_index, current_dir).unwrap();
+
+        $(
+            #[allow(unused)]
+            'block: {
+                let (comp, help) = completions.remove(0);
+                assert_eq!(comp.to_string_lossy(), $comp);
+                $(
+                    assert_eq!(help.unwrap().to_string(), $help, "Expected `{}` to have help message `{}`.", $comp, $help);
+                    break 'block;
+                )?
+                assert!(
+                    help.is_none(),
+                    "Expected `{}` to not have help message `{}`.",
+                    $comp,
+                    help.unwrap()
+                );
+            }
+        )+
+        assert!(completions.is_empty(), "{completions:?}")
+    };
+}
+
+const HELP: &str = "Print this message or the help of the given subcommand(s)";
+const HELP_SHORT: &str = "Print help";
```

Comment:
I feel trying to consolidate this makes it harder to see what the asserts are doing when combined with bespoke macro syntax and makes it harder to evaluate the completions as a whole

Reply:
I'm actually thinking, that it would make more sense, to only do the completion in a helper, i.e.:


```rs
snapshot::assert_eq("something\tHelp for something.",
complete!(cmd, "so[TAB] hello")
```


That way, assertion would be visible, and in case it supports in code updating at some point would be compatible.